### ISSUE TITLE
[2006.09882] Unsupervised Learning of Visual Features by Contrasting Cluster Assignments

### ISSUE THREAD

**Issue by riron1206** [riron1206/read_paper#37]
## 概要
https://arxiv.org/abs/2006.09882
SwAV
画像分類の自己教師あり学習の方法
- SimCLRなどのようにバカでかいバッチサイズ使わずに自己教師あり学習できるし学習も早い
- 画像サイズを224 * 224と96 * 96でマルチクロップした埋め込みベクトルにクラスタリングを適応する
  割り当てたクラスターでマルチクロップのベクトルが同じ予測になるようにlossを計算する
- クラスターの方法にSinkhorn-Knopp(シンクホーン-クノップ)定理というものを使う
- ResNet50でマルチクロップした埋め込みベクトル作成→
  プロトタイプベクトルのスコアに変換（クラスタリングする超空間に投影）→
  シンクホーン定理でクラスターに割りあてる→
  Swap prediction(クロスエントロピーっぽい計算)でloss計算して割り当てたクラスターでマルチクロップのベクトルが同じ予測になるようにする
- クラスターの数はプロトタイプベクトルの数と同じにする必要がある
  プロトタイプベクトルの数はハイパーパラメータ
- Swap predictionは、softmax(プロトタイプベクトルのスコア/温度パラメータ) と softmax(プロトタイプベクトルのスコア/温度パラメータ)にシンクホーン定理適応してた値 とで クロスエントロピーっぽい計算する操作
  各マルチクロップのベクトルを使っているからSwapと名付けてるみたい
- マルチクロップを学習データに入れるのが重要なようでSimCLRや2つのクラスタリングベースのモデル（DeepClusterとSeLa）などの自己監視方式のパフォーマンスも向上させた
  マルチクロップにより画像の数増やせる

![swav](https://camo.githubusercontent.com/422d479e294553eceac31e1defa3ae5084d45cbaa035e5f1f8e9e5eec4bbc87e/68747470733a2f2f646c2e666261697075626c696366696c65732e636f6d2f64656570636c75737465722f616e696d617465642e676966)

![NN](https://i.ibb.co/TtSW4Fd/figure-3.png)

<img src="https://qiita-user-contents.imgix.net/https%3A%2F%2Fi.imgur.com%2FHWAME8F.png?ixlib=rb-1.2.2&auto=format&gif-q=60&q=75&w=1400&fit=max&s=ed69fac03dd3fd3d32f7f98a589942e4" width="500">

※自己教師あり学習(Self-Supervised Learning(SSL))：データ自身から独自のラベルを機械的に作り画像の表現を学ばせるようなタスク。人手のラベル付け不要
　最近のSSLモデルは、Contrastive Learningという手法を用いることで教師あり学習モデルに肩を並べるほどの性能を示している
　https://qiita.com/omiita/items/a7429ec42e4eef4b6a4d
※Contrastive Learning(コンストラスティブラーニング)：「似ているデータは潜在空間でも似た埋め込みベクトルになり、異なるデータは潜在空間でも異なる埋め込みベクトルになる」ということを学習させる
　全結合層に通した画像の埋め込みベクトルの距離を近づける遠ざけるようなメトリックラーニング的な学習方法
　https://qiita.com/omiita/items/a7429ec42e4eef4b6a4d

## 新規性・差分
- クラスター使ったContrastive Learning

## コメント
- 公式のgithub: 
    - https://github.com/facebookresearch/swav (pytorch)

- 非公式のgithub
    - https://github.com/PytorchLightning/pytorch-lightning-bolts (pytorch-lighting)  # これが唯一動かせた
    - https://github.com/keremturgutlu/self_supervised/tree/master/ (pytorch(fast.ai))
    - https://www.kaggle.com/anonamename/swav-tf-simple (TF(処理簡略化していて完全な実装ではない))

<br>

- 参考サイト
    - https://wandb.ai/authors/swav-tf/reports/Unsupervised-Visual-Representation-Learning-with-SwAV--VmlldzoyMjg3Mzg # 非常に詳細に解説してくれてる
    - https://medium.com/syncedreview/facebook-inria-propose-high-performance-self-supervised-technique-for-cv-tasks-39479af4afa9
